evm/db/journal.py

```diff
@@ -0,0 +1,210 @@
+import uuid
+
+from cytoolz import (
+    merge,
+)
+
+from evm.db.backends.base import BaseDB
+from evm.exceptions import ValidationError
+
+
+class Journal(object):
+    """
+    A Journal is an ordered list of checkpoints.  A checkpoint is a dictionary
+    of database keys and values.  The values are the "original" value of that
+    key at the time the checkpoint was created.
+    """
+    checkpoints = None
+
+    def __init__(self):
+        self.checkpoints = []
+        self.journal_data = {}
+
+    @property
+    def latest_id(self):
+        """
+        Returns the checkpoint_id of the latest checkpoint
+        """
+        return self.checkpoints[-1]
+
+    @property
+    def latest(self):
+        """
+        Returns the dictionary of db keys and values for the latest checkpoint.
+        """
+        return self.journal_data[self.latest_id]
+
+    @latest.setter
+    def latest(self, value):
+        """
+        Setter for updating the *latest* checkpoint.
+        """
+        self.journal_data[self.latest_id] = value
+
+    def add(self, key, value):
+        """
+        Adds the given key and value to the latest checkpoint.
+        """
+        if not self.checkpoints:
+            # If no checkpoints exist we don't need to track history.
+            return
+        elif key in self.latest:
+            # If the key is already in the latest checkpoint we should not
+            # overwrite it.
+            return
+        self.latest[key] = value
+
+    def create_checkpoint(self):
+        """
+        Creates a new checkpoint.
+        """
+        checkpoint_id = uuid.uuid4()
+        self.checkpoints.append(checkpoint_id)
```

Comment:
I guess you now use uuids (instead of ints) to identify checkpoints to make it harder to accidentally revert/commit the wrong checkpoint?  Maybe it'd be worth adding a comment to __init__() describing that?